internal/controllers/addon/controller.go

```diff
@@ -335,8 +335,5 @@ func (r *AddonReconciler) Reconcile(
 			return result, nil
 		}
 	}
-
-	// After last phase and if everything is healthy
-	reportReadinessStatus(addon)
```

Comment:
This should be brought back.

Ref: https://github.com/openshift/addon-operator/pull/303/files#r1102396672

Reply:
The way we assume availability in the controller loop is - if there is no error or if retryAfter is not set -> Everything is good(set addon to available).
This is not what we want right, each sub-reconciler has its own part and should be able to set status conditions on its own. For example with the current structure, lets say one of the  sub-reconciler set the available condition to false for a reason which is not exactly an error(which is what I'm doing here, the addon being uninstalled is not an error or something that need to be retried later). The top level controller doesn't take that into consideration, because it has seen no errors, it goes ahead and overwrites the available condition to true.

Reply:
> This is not what we want right, each sub-reconciler is doing its own thing and should be able to set status conditions on its own. For example with the current structure, lets say one of the sub-reconciler set the available condition to false for a reason which is not exactly an error(which is what I'm doing here, the addon being uninstalled is not an error or something that need to be retried later). The top level controller doesn't take that into consideration, because it has seen no errors, it goes ahead and overwrites the available condition to true.

the top-level controller actually takes that into consideration.

Whenever a nil-err and a retry-after happens (result.IsZero() is not nil), the `reportReadinessStatus(addon)` wouldn't execute and [it would straight away return from the main control loop](https://github.com/openshift/addon-operator/blob/5f84edae59a1acfc2e2c1ff246d5f22d0d410759/internal/controllers/addon/controller.go#L334-L336)

But that's not just it,
the control loop would ultimately, execute the `defer` section which would finally report the addon status set by the individual sub-reconcilers - https://github.com/openshift/addon-operator/blob/5f84edae59a1acfc2e2c1ff246d5f22d0d410759/internal/controllers/addon/controller.go#L286-L288

Reply:
But I'm talking about situations where there is no error and `result` is zero.

Reply:
a "no error and zero-result" is a sub-reconcilers way of saying "everything is good"

a "no error and a non-zero-result" gives the sub-reconciler the opportunity to set a custom status of the addon with something wrong/not-ideal.

a "non-nil error" gives the sub-reconciler the opportunity to say "some error happened so don't update status's as it might be wrong/stale because of error'ed reconciliation"

Reply:
Do we happen to have a situation where we want to report a custom "successful" condition from a sub-reconciler? if that's the case, then I guess that's redundant with respect to the main control loop's `reportReadinessAddon` function, and we should only design our status reporting to be compliant with reporting negative status's and letting the main control loop presume everything to be OK upon seeing no negative statuses.